### ISSUE TITLE
Treat Weights as Constants When Inferencing

### ISSUE THREAD

**Issue by yaochengji** [pytorch/xla#3081]
## 🚀 Feature
When using torch/xla for inferencing, could we treat weights as constants? If it is done, there will be more chance for XLA to optimize the graph better, Ex. fusing BN into Conv.

**Comment by JackCaoG** [comment]
I think the challenge here is that when weights changes we will have to recompile the graph. That's why we only make 0 and 1 to be constant and leave the rest of them to be the device data.

**Comment by yaochengji** [comment]
We could let user control this. There could be a flag named `TREAT_WEIGHTS_AS_CONSTANTS` and if it is set to `1` by user, then XLA will consider all weights as constants.

**Comment by yaochengji** [comment]
> We could let user control this. There could be a flag named `TREAT_WEIGHTS_AS_CONSTANTS` and if it is set to `1` by user, then XLA will consider all weights as constants.

Users should be responsible for that the weights will never change.

**Comment by JackCaoG** [comment]
It should be couple lines of change in [here](https://github.com/pytorch/xla/blob/master/torch_xla/csrc/tensor.cpp#L244), but then you need to separate weights from other scalar. Do you mind testing it out the speed? This is a bit risky(many more recompilation if using wrong) and a power user feature. Would be nice to find out what kind of speed gain we are talking about.

**Comment by yaochengji** [comment]
Thanks, I will have a try.

**Comment by JackCaoG** [comment]
@cjolivier01 I vaguely remember you had similar feature request and open a pr a while ago, is that the case?

**Comment by cjolivier01** [comment]
I did have a PR for "treat all scalars as constants". 

I am gathering that the ask is to extend this to weights (which may not be scalars) and hoping they will get constant-folded during the HLO passes?  My first inclination is that this is a bit trickier because it's not always so easy to recognize what is a weight in a general way (as it is a scalar), especially when it isn't going to be an alias input/output from this low of a level in the current code path (although I could be wrong).  

Off the top of my head, I can think of a couple of ways to get around this, but neither of them are very pretty.  There is some knowledge up in aten about what is and isn't a weight (i.e. api params, at::Tensor is_leaf(), requires_grad()) -- assuming requires_grad() holds for inference.  Since TorchScript is mostly for inference, I wonder if they do this sort of optimization (although they don't have XLA HLO passes, of course) and if so, by what mechanism do they make this distinction at compile-time?  @ailzhang

**Comment by cjolivier01** [comment]
I apologize -- I take that back.  DeviceData nodes will tend to be weights, in which case it wouldn' t be too difficult.  I recall now that the distinguish problem I had been having was due to something specific to my use-case at the time (pruning some superfluous updates which weren't actually model parameters).

**Comment by cjolivier01** [comment]
btw here was that PR: https://github.com/pytorch/xla/pull/2543

**Comment by yaochengji** [comment]
@JackCaoG Here's the PR: https://github.com/pytorch/xla/pull/3082

I did a rough test on Resnet (`print(loss.cpu())` on every iteration) and find that the performance gain is negligible.

**Comment by JackCaoG** [comment]
@yaochengji Thanks for testing out, in that case I would prefer not to merge this pr. It doesn't provide enough speed up and can cause a lot of trouble when being used incorrectly.

**Comment by yaochengji** [comment]
@JackCaoG I'll close the PR for now. After finding a more suitable situation, we could reconsider it.

**Comment by JackCaoG** [comment]
Thanks @yaochengji , I will close this issue for now. Feel free to reopen if you have more topic to discuss.
